Add UpdateHub.io support lib

otavio opened the pull request:
UpdateHub provides an end-to-end solution for Over-The-Air update
of embedded devices.

Signed-off-by: Christian Tavares <christian.tavares@ossystems.com.br>
Reviewed-by: Otavio Salvador <otavio@ossystems.com.br>

Commit:
lib: updatehub: Add UpdateHub.io support

UpdateHub is an enterprise-grade solution which makes simple to
remotely update all your embedded devices in the field. It
handles all aspects related to sending Firmware Over-the-Air(FOTA)
updates with maximum security and efficiency, while you focus in
adding value to your product.

Signed-off-by: Christian Tavares <christian.tavares@ossystems.com.br>
Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>

pfalcon commented:
This looks pretty cool, but I doubt it belongs to subsys/ . Rather, lib/ . I also would suggest to join initiative of https://github.com/zephyrproject-rtos/zephyr/issues/12129 and put headers under include/zephyr/ right away. But don't haste with these changes, until others provided feedback too.

But you definitely would need to act on @zephyrbot's comments ;-).

pfalcon commented:
Also, please consider how would we ensure that this works over time. We'd need a sample, tests, etc.

carlescufi commented:
Thank you for this contribution!
This looks pretty good already. I agree that with @pfalcon this does not belong directly on `subsys/`. I think the reasonable options are:

- `lib/dfu/updatehub`
- `lib/updatehub`

Also please include a buildable test (a simple application that uses the library) in `tests/`. This will allow us to include the new code in CI so that we make sure it is not broken in the future.

Finally we would need a bit of documentation in `doc/`.

Laczen reviewed (changes requested):
Hi @chtavares592 and @otavio, thanks for this PR. My review is only limited to the use of nvs; as my knowledge on the other (most important part) is limited:
You are using NVS to store the id, in the creation of the nvs filesystem you are using a very short (64 bytes) sector size. The sector size should be a multiple of the flash_erase_page size, so probably 64 is to small. You are also using FLASH_AREA_STORAGE as the offset of the nvs file system, this would make it impossible for other modules to use it.
My advice would be to use the settings module to store the id, this will allow other modules to use the same storage area. If you need advice/help on how to do this, let me know.

nvlsianpu commented on subsys/updatehub/Kconfig:
```diff
@@ -0,0 +1,106 @@
+	select NVS
+	default "n"
+	help
+		The options below ought to show only if UpdateHub support is enabled.
```
help should explain what the key UPDATEHUB does.

otavio commented:
@pfalcon 
> This looks pretty cool, but I doubt it belongs to subsys/ . Rather, lib/ . I also would suggest to join initiative of #12129 and put headers under include/zephyr/ right away. But don't haste with these changes, until others provided feedback too.

We can do the inclusion of the headers on the new location for sure; we also are in process of moving it to `lib/updatehub` as suggested by @carlescufi.

@Laczen 
> You are using NVS to store the id, in the creation of the nvs filesystem you are using a very short (64 bytes) sector size. The sector size should be a multiple of the flash_erase_page size, so probably 64 is to small. You are also using FLASH_AREA_STORAGE as the offset of the nvs file system, this would make it impossible for other modules to use it.
> My advice would be to use the settings module to store the id, this will allow other modules to use the same storage area. If you need advice/help on how to do this, let me know.

We are looking at this and hope to have something to share soon (or questions, hehe)...

Laczen commented:
> @Laczen
> 
> > You are using NVS to store the id, in the creation of the nvs filesystem you are using a very short (64 bytes) sector size. The sector size should be a multiple of the flash_erase_page size, so probably 64 is to small. You are also using FLASH_AREA_STORAGE as the offset of the nvs file system, this would make it impossible for other modules to use it.
> > My advice would be to use the settings module to store the id, this will allow other modules to use the same storage area. If you need advice/help on how to do this, let me know.
> 
> We are looking at this and hope to have something to share soon (or questions, hehe)...

OK, a example on how it is used can be found in:
`zephyr/samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c`

otavio commented:
We did get the settings integrated by then we found a problem: we lose the information when upgrade is done.

Essentially, what seems to be happening is:

```
Step 1: updatehub -> set updatehub/ip -> reboot
Step 2: mcuboot -> swap slots -> reboot
Step 3: updatehub -> get updatehub/ip (fail!)
```

@Laczen any suggestion how to approach this?

mike-scott commented:
@otavio Thank you for the PR.  Overall I agree with the others in several regards (location in lib/, PR quality is good and the need for a sample / test).

I notice you are using device_identity.c which has some hard-coded values per SoC for reading device unique values.  There is a Zephyr-generic solution mostly in place here: https://github.com/zephyrproject-rtos/zephyr/tree/master/drivers/hwinfo  And I believe NXP support is nearly approved.  You might consider using that API for getting the initial seed for your unique identifier.

RE: the settings values going away after update, make sure the partition sizes aren't overlapping between MCUboot, slot0, slot1, mcuboot's scratch area, the NVS settings, etc.  If they do overlap it could explain why those settings go away during OTA.

Laczen commented:
I agree with @mike-scott , probably there is some problem with the flash partitions. If you can put your sample somewhere I can have a look (please don't forget to mention your board).

mike-scott commented:
@otavio could provide a diff of the previous commit vs. the new one pushed a few days ago?

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+for updates and installing them without requiring user intervention. You
+can access the sample source code for this mode `updatehub_polling`_.
+
+.. _updatehub_polling: updatehub_polling
```
I'm guessing you'll provide a real external link to where this update polling content is?  As written this won't work.

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+update now or later. You can access the sample source code for this mode
+`updatehub_manual`_.
+
+.. _updatehub_manual: updatehub_manual
```
Same here.  This needs a link to the manual, something like:
```
.. _updatehub_manual: https://somewebsite.com/manualdoc.html
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+    :build-dir: fdrm_k64f_build
+    :goals: build
+
+.. updatehub_sample_sign:
```
```suggestion
.. _updatehub_sample_sign:
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+.. code-block:: console
+
+    sudo pyocd-flashtool <path-to-signed.bin>
```
``(bin|hex)`` to be consistent?

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 6: Signing the test image
+================================
+
+For the update to be correctly validated on the server you will need sign the
```
```suggestion
For the update to be correctly validated on the server, you must sign the
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+
+Step 7: Create a package with UpdateHub Utilities (``uhu``)
+=======================================================
```
The title underline must be at least as long as the title above it, so add a few more ``===``'s

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 7: Create a package with UpdateHub Utilities (``uhu``)
+=======================================================
+
+First you need install UpdateHub Utilities (``uhu``) at your system, using:
```
```suggestion
First, install UpdateHub Utilities (``uhu``) on your system, using:
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+.. code-block:: console
+
+    pip3 install uhu
```
will this need ``sudo`` to install, or should you use ``--user``?

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+    uhu product use "e4d37cfe6ec48a2d069cc0bbb8b078677e9a0d8df3a027c4d8ea131130c4265f"
+
+Also add the package and it's mode (``zephyr``):
```
```suggestion
Then, add the package and its mode (``zephyr``):
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 8: Start the updatehub-ce-server
+=====================================
+
+For default the updatehub application is set to start on the official server.
```
```suggestion
By default, the updatehub application is set to start on the official server.
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+=====================================
+
+For default the updatehub application is set to start on the official server.
+For more details on how to use the official server please reffer to the
```
```suggestion
For more details on how to use the official server please refer to the
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+For default the updatehub application is set to start on the official server.
+For more details on how to use the official server please reffer to the
+documentation on the `updatehub.io`_.
```
```suggestion
documentation on `updatehub.io`_.
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+.. _updatehub.io: https://updatehub.io
+
+However if you would like to use your own server. The steps bellow explain how
```
```suggestion
If you would like to use your own server, the steps below explain how
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+.. _updatehub.io: https://updatehub.io
+
+However if you would like to use your own server. The steps bellow explain how
+updatehub works with updatehub-ce-server.
```
```suggestion
updatehub works with updatehub-ce-server
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+However if you would like to use your own server. The steps bellow explain how
+updatehub works with updatehub-ce-server.
+Starting updatehub-ce-server you can just running the following Docker command:
```
```suggestion
running, started by the following Docker command:
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+    docker run -d -p 8080:8080 -p 5683:5683/udp --rm  updatehub/updatehub-ce-server:latest
+
+Using this server you can set the variable ``CONFIG_UPDATEHUB_PROBE_SERVER``
```
```suggestion
Using this server you can set the option :option:`CONFIG_UPDATEHUB_PROBE_SERVER`
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+Using this server you can set the variable ``CONFIG_UPDATEHUB_PROBE_SERVER``
+on the ``prj.conf`` file with your local ip address or you can use the ip address
+directly on the shell command line, these steps bellow explain using the shell
```
```suggestion
directly on the shell command line.  The steps below explain using the shell
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+directly on the shell command line, these steps bellow explain using the shell
+command line.
+
+Step 9: Add the package the server
```
```suggestion
Step 9: Add the package to the server
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+command line.
+
+Step 9: Add the package the server
+==================================
```
and adjust the title underline

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 9: Add the package the server
+==================================
+
+Now you need add the package at updatehub-ce-server, for this you will need to
```
```suggestion
Now, add the package to the updatehub-ce-server by
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+==================================
+
+Now you need add the package at updatehub-ce-server, for this you will need to
+open your browser to the server URL, ``<your-ip-addres>:8080``, and then log
```
```suggestion
opening your browser to the server URL, ``<your-ip-addres>:8080``, and logging
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+Now you need add the package at updatehub-ce-server, for this you will need to
+open your browser to the server URL, ``<your-ip-addres>:8080``, and then log
+into the server using ``admin`` as the login and password by default.  After
```
is the password "password"?

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Now you need add the package at updatehub-ce-server, for this you will need to
+open your browser to the server URL, ``<your-ip-addres>:8080``, and then log
+into the server using ``admin`` as the login and password by default.  After
+log-in you must click on the package menu and then ``UPLOAD PACKAGE``, and select
```
```suggestion
logging in, click on the package menu, then ``UPLOAD PACKAGE``, and select
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+open your browser to the server URL, ``<your-ip-addres>:8080``, and then log
+into the server using ``admin`` as the login and password by default.  After
+log-in you must click on the package menu and then ``UPLOAD PACKAGE``, and select
+the package built on the step 7.
```
```suggestion
the package built in step 7.
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+log-in you must click on the package menu and then ``UPLOAD PACKAGE``, and select
+the package built on the step 7.
+
+Step 10: Registry device on server
```
```suggestion
Step 10: Register device on server
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 10: Registry device on server
+==================================
+
+For registry your device at updatehub_ce_server you need open your terminal
```
```suggestion
Register your device at updatehub_ce_server by using a terminal
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+==================================
+
+For registry your device at updatehub_ce_server you need open your terminal
+where you use for debbuging the board, and type the following command:
```
```suggestion
session on the system where you were debugging the board, and type the following command:
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+    updatehub run <your-local-ip>
+
+If everything is alright will print on the screen ``No update available``.
```
```suggestion
If everything is alright, it will print on the screen ``No update available``.
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 11: Create a rollout
+=========================
+
+You need go where the updatehub_ce_server is open on the browser and click on
```
Change to:
```
In the browser where the updatehub_ce_server is open, click on
``menu Rollout`` and then ``CREATE ROLLOUT``. Select the version of the package
that you added in step 9.  With that, the update is published, and the server
is ready to accept update requests.
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 12: Run the update
+=======================
+
+Open your terminal that you use for debbuging the board, and type the following
```
```suggestion
Back in the terminal session that you used for debugging the board, type the following
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+    updatehub run <your-local-ip>
+
+And then wait. The board will ping the server, check if there's any new updates,
```
```suggestion
And then wait. The board will ping the server, check if there are any new updates,
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+And then wait. The board will ping the server, check if there's any new updates,
+and then download the update package you've just created. If everything goes
+fine the message `Image flashed successfully, you can reboot now` will be
```
```suggestion
fine the message ``Image flashed successfully, you can reboot now`` will be
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+Step 13: Reboot the system
+==========================
+
+Open your terminal where you use for debbuging the board, and type the following
```
```suggestion
In the terminal you used for debugging the board, type the following
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+    kernel reboot cold
+
+Your board will reboot and then start with the new image. After rebooting the
+board will automatically ping the server again and the message `No update
```
```suggestion
board will automatically ping the server again and the message ``No update
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,246 @@
+
+Your board will reboot and then start with the new image. After rebooting the
+board will automatically ping the server again and the message `No update
+available` will be printed on the terminal.
```
```suggestion
available`` will be printed on the terminal.
```

dbkinder reviewed (changes requested):
Please review my clarity edits that I didn't introduce a technical error.

otavio commented:
@mike-scott 
> @otavio could provide a diff of the previous commit vs. the new one pushed a few days ago?

I am afraid we can't. We amended the changes and we later dropped those branches. We are now working on fixing the `mcuboot` image swap issue and we ought to have it ready for a new review.

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,239 @@
+update all your Linux-based devices in the field. It handles all aspects
+related to sending Firmware Over-the-Air (FOTA) updates with maximum
+security and efficiency, while you focus in adding value to your product.
+These are samples on how to implement simple, polling or manual, local update
```
Delete this sentence (it's already said on the next line).

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,239 @@
+Overview
+********
+
+UpdateHub is an enterprise-grade solution which makes simple to remotely
```
```suggestion
UpdateHub is an enterprise-grade solution which makes it simple to remotely
```

dbkinder commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,239 @@
+UpdateHub is an enterprise-grade solution which makes simple to remotely
+update all your Linux-based devices in the field. It handles all aspects
+related to sending Firmware Over-the-Air (FOTA) updates with maximum
+security and efficiency, while you focus in adding value to your product.
```
```suggestion
security and efficiency, while you focus on adding value to your product.
```

dbkinder reviewed (changes requested):
a few more doc edits...

dbkinder reviewed (approved):
+1 for docs, thanks!

mike-scott reviewed (commented):
Couple of questions:

- Do we feel like samples/updatehub is the right location for the sample?  I understand that there might be some discussion as to whether this should be samples/net/* or maybe somewhere else?  But, primarily this is a networking sample.
- There are 2 samples complete with Kconfig and prj.conf files in the PR.  When I compare them, they don't seem to be that different.  Perhaps a single sample with Kconfig option to enable the polling mode vs. manual update mode?  This will probably be easier to maintain over time.
- There is also very little discussion of security in this sample.  I understand that this may be something that UpdateHub.io would plan for enterprise users, however, security should be something that's presented by default and not "added on later".  I would feel a lot better about the sample if it had options for CoAPS/DTLS support.

mike-scott commented:
Also, the following in the documentation isn't necessarily true:
```
* The Zephyr port of ``UpdateHub`` is configured to run on a Freedom-K64F
  MCU by default. The application should build and run for other platforms
  without modification.
```

There are only a handful of devices which use ethernet or have a direct internet connection such as an LTE-M modem.   Overlay files would be needed to support BLE 6lowpan, 802.15.4 or OpenThread configurations as well as the understanding that most other connectivity options would require an edge gateway of some sort (Border Router, etc).

mike-scott commented on samples/updatehub/README.rst:
```diff
@@ -0,0 +1,237 @@
+    sudo pyocd-flashtool <path-to-signed.bin>
+
+
+Step 8: Signing the test image
```
Docs seem to be missing Step 7?

mike-scott reviewed (changes requested):
docs seem to be missing a step which sets the signing key for the package.  When running the end of step 9: `uhu package archive --output <name-of-package>.pkg` I'm presented the following error:
```
Error: Could not find any private key.
```

otavio commented:
@mike-scott we are finishing the rework on the items you mentioned. Thank you for your detailed review. 

> There is also very little discussion of security in this sample.  I understand that this may be
> something that UpdateHub.io would plan for enterprise users, however, security should be 
> something that's presented by default and not "added on later".  I would feel a lot better about
> the sample if it had options for CoAPS/DTLS support.

This is being worked on. As you can see at https://github.com/go-ocf/go-coap/issues/23 the `go-coap` does not yet support DTLS.

We are working on adding support for DTLS for `go-coap` so we can implement it on our server components and avoid the overhead of the TCP/IP for the MCU as this would increase the resource use of the MCU memory considerably and we are trying very hard to be as lean as possible.

That said, we'd like to ask if it would be a blocker for first version merging or not?

otavio commented:
Folks, besides the DTLS support (which we are working on), is there anything required to be changed prior merging?

otavio commented:
@mike-scott 

> docs seem to be missing a step which sets the signing key for the package. When running the end of step 9: `uhu package archive --output <name-of-package>.pkg` I'm presented the following error:
> 
> ```
> Error: Could not find any private key.
> ```

We removed this requirement as for Zephyr it does not make much sense due to the image already be signed and validated by the `mcuboot`. So we did a new release without this requirement.

Laczen commented:
@otavio, I have a PR to change the settings module. This will require some minor changes to your PR.

otavio commented:
> @otavio, I have a PR to change the settings module. This will require some minor changes to your PR.

Which PR? we can start updating it right away.

Laczen commented:
> > @otavio, I have a PR to change the settings module. This will require some minor changes to your PR.
> 
> Which PR? we can start updating it right away.

PR 12998, but wait to update, your PR might be faster.

mike-scott commented:
@otavio couple of comments:

1. I won't block this for lack of DTLS support if you say it's in-progress.
2. The diff for the current PR is quite large compared to the last one.  Normally, reviewers should be given a warning of some sort to indicate that a large portion of the content has changed.  Otherwise, they might assume that only the items pointed out were changed and unknowingly approve of several other un-reviewed changes.
3. new commit message mentions "remotely update all your Linux-based devices", perhaps this shouldn't focus on Linux-based as this would be a patch for Zephyr.
4. Movement from samples/updatehub -> samples/net looks good (with CODEOWNER changes to match)
4. You should remove the "depends on UPDATEHUB" and "depends on UPDATEHUB_CE_SERVER" in lib/updatehub/Kconfig now that there are protected blocks of:
"if UPDATEHUB" and "if UPDATEHUB_CE_SERVER" (@ulfalizer you can thank me later)

I will need some time to re-review the rest of the changes.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+			goto cleanup;
+		}
+
+		int i;
```
Variable declarations to the beginning of the block

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+			strcat(&sha256_image_dowloaded[i], buffer);
+		}
+
+		if (strcmp(sha256_image_dowloaded, _update_info.sha256sum_image) !=
```
The indentation looks a bit ugly here, what about splitting after the first parameter

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+	flash_img_init(&_ctx.flash_ctx);
+
+	_ctx.downloaded_size = 0;
+	int verification_download = 0;
```
vars again

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+		goto error;
+	}
+
+	struct report report;
```
ditto

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+		break;
+	default:
+		report.previous_state = "";
+	}
```
Does not make difference here but better to put break; after all the cases. There are some misra checks that will complain about these.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+		goto cleanup;
+	}
+
+	struct resp_probe_some_boards metadata_some_boards;
```
vars

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+
+	LOG_INF("Image flashed successfully, you can reboot now");
+
+#if defined(CONFIG_UPDATEHUB_LOG)
```
What is the purpose of this while loop

chtavares592 replied:
Just wait for the LOG_INF print on the screen.

jukkar replied:
You do not need to do it like this. If you want synchronous prints, then just enable CONFIG_LOG_IMMEDIATE

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+	case UPDATEHUB_UNCONFIRMED_IMAGE:
+		LOG_ERR("Image is unconfirmed. Rebooting to revert back to previous"
+			"confirmed image.");
+#if defined(CONFIG_UPDATEHUB_LOG)
```
same here

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+			k_sleep(5);
+		}
+#endif
+		sys_reboot(0);
```
Instead of hard coded value, please use proper symbol here (SYS_REBOOT_COLD vs SYS_REBOOT_WARM)

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+	k_delayed_work_submit(&work, 0);
+}
+
+#if defined(CONFIG_UPDATEHUB_SHELL)
```
This source file is quite large. Perhaps the shell support could be placed into another .c file.

jukkar commented on samples/net/updatehub/Kconfig:
```diff
@@ -0,0 +1,30 @@
+#
+mainmenu "Sample app Configuration"
+
+config ZEPHYR_BASE
```
I think this ZEPHYR_BASE setting is no longer needed

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+********
+
+UpdateHub is an enterprise-grade solution which makes it simple to remotely
+update all your Linux-based devices in the field. It handles all aspects
```
s/Linux/Zephyr/

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+Overview
+********
+
+UpdateHub is an enterprise-grade solution which makes it simple to remotely
```
Could we have a link to specification of updatehub here?

There could be a link and chapter about updatehub in /doc/reference/networking/protocols.rst

otavio replied:
What do you mean by specification here?

jukkar replied:
> What do you mean by specification here?

For example LWM2M documentation has links to relevant specifications, see example here https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/14180/docs/reference/networking/lwm2m.html#overview
I was wondering that if someone is reading this sample documentation via browser, he/she could just follow link to some webpage that could explain things about updatehub. Even wikipage etc. would be useful.

otavio replied:
We have https://docs.updatehub.io/; this needs to have a Zephyr specific session added but seems to be the most logical place to point the user to.

jukkar replied:
yes, that looks exactly what I meant

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+Caveats
+*******
+
+* The Zephyr port of ``UpdateHub`` is configured to run on a Freedom-K64F
```
Please add a :ref: to frdm-k64f board, so something like this
```
:ref:`Freedom-K64F <frdm_k64f>`
```

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+For more details on how to use the official server please refer to the
+documentation on `updatehub.io`_.
+
+.. _updatehub.io: https://updatehub.io
```
You could collect these links at the end of the document. It is easier to read the file if they are not cluttering the text.

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+
+    docker run -d -p 8080:8080 -p 5683:5683/udp --rm  updatehub/updatehub-ce-server:latest
+
+Using this server you can set the option :option: ``CONFIG_UPDATEHUB_SERVER``
```
Remove the space after :option:

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+
+Using this server you can set the option :option: ``CONFIG_UPDATEHUB_SERVER``
+on the ``prj.conf`` file with your local ip address and the option :option:
+``CONFIG_UPDATEHUB_CE_SERVER`` with true.
```
ditto

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+``UpdateHub`` can be built for the fdrm_k64f as follows:
+
+.. zephyr-app-commands::
+    :zephyr-app: samples/updatehub/updatehub_<manual-or-polling>
```
samples directory wrong

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,240 @@
+.. zephyr-app-commands::
+    :zephyr-app: samples/updatehub/updatehub_<manual-or-polling>
+    :board: fdrm_k64f
+    :build-dir: fdrm_k64f_build
```
usually the build-dir can be omitted so that default one is used

jukkar commented on samples/net/updatehub/prj.conf:
```diff
@@ -0,0 +1,25 @@
+CONFIG_UPDATEHUB=y
+
+#Example of product_uid, you must set you product uid.
+CONFIG_UPDATEHUB_PRODUCT_UID="e4d37cfe6ec48a2d069cc0bbb8b078677e9a0d8df3a027c4d8ea131130c4265f"
```
In order to avoid the user to modify this file if he changes the product uid, this setting could be placed in a separate overlay config file. This way this config file would only contain generic settings which are not related to user environment.

The CONFIG_UPDATEHUB_SERVER="<your-addres-ip>" could be in the same overlay file.

jukkar commented on samples/net/updatehub/prj.conf:
```diff
@@ -0,0 +1,25 @@
+
+# Debug helpers
+CONFIG_UPDATEHUB_LOG=y
+CONFIG_UPDATEHUB_LOG_LEVEL=3
```
You need to use the log level template file, after that this setting would be called CONFIG_UPDATEHUB_LOG_LEVEL_INF=y

jukkar commented on CODEOWNERS:
```diff
@@ -206,13 +206,15 @@
 /kernel/device.c                          @andrewboie @andyross @nashif
 /kernel/idle.c                            @andrewboie @andyross @nashif
 /samples/basic/servo_motor/*microbit*     @jhe
+/lib/updatehub/                           @chtavares592 @otavio
```
What about placing the updatehub into `subsys/net/lib` directory?

otavio replied:
As discussed in previous comments, we are open to moving it to somewhere else when more OTA stuff gets in, but it is not limited to net use, so we prefer to not move it now.

jukkar replied:
> but it is not limited to net use, so we prefer to not move it now.

Sure, np

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+#endif
+
+static struct updatehub_context {
+	struct net_context *net_ctx;
```
The net_ctx looks like a leftover from some earlier version.

otavio commented on include/updatehub.h:
```diff
@@ -0,0 +1,75 @@
+
+#ifndef _UPDATEHUB_H_
+#define _UPDATEHUB_H_
+
```
> I am just pondering that should the include file be placed into `include/net`, what do you think?

I am not sure I buy the net change. Mainly because in future we could have other ways of putting a new firmware on the device, for example using a pendrive (for the devices which has the connector). 

So I'd prefer to not limit to the net namespace.

jukkar replied:
Fair enough, I am fine if it is not part of network subsystem.

otavio commented on lib/updatehub/Kconfig:
```diff
@@ -0,0 +1,105 @@
+		handles all aspects related to sending Firmware Over-the-Air
+		(FOTA) updates with maximum security and efficiency, while you
+		focus in adding value to your product.
+
```
What is the best practice here? Should we put the `if SOMETHING` or just `depends on SOMETHING`?

pfalcon commented:
> I am not sure I buy the net change. Mainly because in future we could have other ways of putting a new firmware on the device, for example using a pendrive (for the devices which has the connector).

+1 to such an approach. I believe there was a similar comment from @mike-scott previously (re: getting this under something-net), sorry if I mix up something. But IMHO, this shouldn't be related to net directly, but rather to some "fwupdate" lib/subsys. For example, any additional infra related to mcuboot would also belong to it, and more may come in the future. They may be OTA or non-OTA, net or non-net (e.g., why may imagine OTA update using SMS, why not, lol).

Note that I don't call to put this under lib/fwupdate/... right away, but we should be ready to do that later. But my point is that we definitely shouldn't put this under subsys/net/ or similar.

otavio commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+		goto error;
+	}
+
+	if (info && info->ai_family == AF_INET) {
```
We can certainly add support for it on the Zephyr agent, but could we do it as a subsequent change?

otavio replied:
As we discussed, once we finish the DTLS support we can work on this.

otavio commented on lib/updatehub/Kconfig:
```diff
@@ -28,78 +26,54 @@ config UPDATEHUB
+module-dep = LOG
+module-str = Log level for BSD sockets compatible API calls
+module-help = Enables logging for sockets code.
+source "subsys/net/Kconfig.template.log_config.net"
```
Please add a new line here.

chtavares592 commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+			enum updatehub_uri_path type)
+{
+	struct coap_packet request_packet;
+	u8_t *data = k_malloc(MAX_PAYLOAD_SIZE);
```
We need to use k_malloc () because the server sends data that can not be overwritten, so as security we are using K_malloc, do you know any idea of another solution that we can use?

otavio replied:
Do you have any similar situation, which has been reworked to avoid the use of malloc, so we can take a look and see if we can come up with an alternative solution?

jukkar replied:
One wild idea is to utilize net_pkt api, as that is available because you are using networking in this system. But perhaps for you it is easier just to stick with k_malloc() in this case.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -425,19 +380,15 @@ static void install_update_cb(void)
 
-		if (strcmp(sha256_image_dowloaded, _update_info.sha256sum_image) !=
-		    0) {
+		if (strcmp(sha256_image_dowloaded,
```
I missed this last time. Please use strncmp() instead here.

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -77,19 +75,19 @@ following Docker command:
-on the ``prj.conf`` file with your local ip address and the option :option:
-``CONFIG_UPDATEHUB_CE_SERVER`` with true.
+Using this server you can set the option :option:``CONFIG_UPDATEHUB_SERVER``
+on the ``prj.conf`` file with your local ip address and the option
```
Please instruct user to create overlay-prj.conf file here, and place all the bits and pieces related to his setup there.

jukkar commented on samples/net/updatehub/overlay-prj.conf:
```diff
@@ -0,0 +1,8 @@
+#Example of product_uid, you must set you product uid.
```
It would be better that this overlay-prj.conf is called overlay-prj.conf.example, and document that user needs to create his own file and use that. This way the overlay-prj.conf is not placed into zephyr version control which could conflict user setup when he updates his tree.

jukkar reviewed (changes requested):
Getting there, but there are still some issues left.

chtavares592 commented on include/updatehub.h:
```diff
@@ -4,9 +4,11 @@
+ * @brief UpdateHub Firmware Over-the-Air for Zephyr Project.
+ * @defgroup updatehub UpdateHub Firmware Over-the-Air
+ * @ingroup lib
+ * @{
```
In the end of file, it has the end group.

otavio commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,48 @@
+
+	for (i = 0; i < length; i++) {
+		snprintk(buf, DEVICE_ID_SIZE, "%02x", hwinfo_id[i]);
+		strncat(id, buf, strlen(buf));
```
@jukkar, we were reviewing it here and in fact, the `hwinfo_get_device_id` already handles the max size of the buffer and it guarantees to return up to it.

So, in line 29, we pass the buffer size. In line 36, we walk on it using the returned size.

I think it is safe ...

mike-scott replied:
Actually, this function assumes `id` was created with enough room for `DEVICE_ID_SIZE`.  I agree with @jukkar  that this function should probably take a 2nd parameter for length of `id` buffer.  And that can be used here in `strncat`.

otavio commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,920 @@
+		hr_family = "IPv4";
+		addr = &net_sin(&info->ai_addr)->sin_addr;
+		buffer = net_addr_ntop(info->ai_family, addr, hr_addr, sizeof(hr_addr));
+		memcpy(user_ip, buffer, strlen(buffer));
```
I am sorry but we did not understand what you mean here. Do you mind to detail it a little bit?

jukkar replied:
I was thinking something like this:
```
struct user_data {
	int user_ip_len;
	char *user_ip;
};

static void dns_cb(enum dns_resolve_status status, struct dns_addrinfo *info,
		   void *user_data)
{
	struct user_data *ud = user_data;
....
	buffer = net_addr_ntop(info->ai_family, addr, hr_addr, sizeof(hr_addr));
	buf_len = strlen(buffer);
	memcpy(ud->user_ip, buffer, MIN(ud->user_ip_len, buf_len));
...
}

static bool start_coap_client(void)
{
	static struct user_data ud;
	static char ip[MAX_IP_SIZE];

	ud.user_ip = ip;
	ud.user_ip_len = sizeof(ip) - 1;

	if (dns_get_addr_info(UPDATEHUB_SERVER, DNS_QUERY_TYPE_A, &dns_id,
			      dns_cb, &ud, NETWORK_TIMEOUT) < 0) {
		LOG_ERR("Could not resolve dns");
		return false;
	}
....
}
```

Could be overkill here if the DNS cb is resolved from only one place. In that case the simplest case would be just to do
```
	buf_len = strlen(buffer);
	memcpy(user_ip, buffer, MIN(MAX_IP_SIZE, buf_len));
```

chtavares592 replied:
Good, I understood now.  We will be using the second option because that happens in just one local. Thanks, @jukkar!!

otavio commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,799 @@
+	struct coap_packet request_packet;
+	u8_t *data = k_malloc(MAX_PAYLOAD_SIZE);
+	int ret = -1;
+	const u8_t *api_header =
```
@chtavares592  please define a constant inside our private .h as UPDATEHUB_API_HEADER; this makes the code easier to read.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,799 @@
+	k_sem_init(&_ctx.semaphore, 0, 1);
+
+	memset(&ip, 0, MAX_IP_SIZE);
+	if (dns_get_addr_info(UPDATEHUB_SERVER, DNS_QUERY_TYPE_A, &dns_id,
```
BTW, the dns resolver function is async one. The function will return immediately, you need to add waiting of the results by yourself here.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,822 @@
+			enum updatehub_uri_path type)
+{
+	struct coap_packet request_packet;
+	int ret = -1;
```
Useless initialization as you are setting the ret variable a value below.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,822 @@
+		goto error;
+	}
+
+	const char *exec = state_name(state);
```
Please place the variable declarations at the beginning of the block.

jukkar commented on samples/net/updatehub/Kconfig:
```diff
@@ -0,0 +1,30 @@
+
+config ZEPHYR_BASE
+	string
+	option env="ZEPHYR_BASE"
```
As I mentioned earlier, you do not need to set the ZEPHYR_BASE any more.

jukkar commented on samples/net/updatehub/Kconfig:
```diff
@@ -0,0 +1,30 @@
+
+config UPDATEHUB_MANUAL
+	bool "UpdateHub manual mode"
+	default n
```
The "default n" is automatically the default, so just remove this line.

jukkar commented on samples/net/updatehub/Kconfig:
```diff
@@ -0,0 +1,30 @@
+	bool "UpdateHub manual mode"
+	default n
+	help
+		Manual mode requires the user to call the server probe and then, if
```
Fix the indentation of help text

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -0,0 +1,242 @@
+    docker run -d -p 8080:8080 -p 5683:5683/udp --rm  updatehub/updatehub-ce-server:latest
+
+Using this server the user need create own ``overaly-prj.conf`` setting the option
+:option:``CONFIG_UPDATEHUB_SERVER`` with your local ip address and the option
```
The :option: links use only one back ticks, search and replace all in this doc.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,817 @@
+		return false;
+	}
+
+	k_sem_take(&ctx.semaphore, NETWORK_TIMEOUT);
```
The NETWORK_TIMEOUT here is the same in `dns_get_addr_info()` few lines above. As the DNS resolving is done in separate thread than this one, we cannot know which one will trigger first. This means that the semaphore might trigger with timeout before the DNS handler thread has called the callback and released the semaphore. Probably the easiest fix is to make this timeout here bigger than the DNS one, for example `NETWORK_TIMEOUT * 2` or similar.

jukkar reviewed (changes requested):
LGTM, just one issue left from my side.

otavio commented:
@jukkar thanks for all your patience and good work; we look forward to get this merged soon :+1:

pfalcon commented:
@otavio, Just to set the expectations right, Zephyr is in the code freeze stage now for the 1.14 release, with officially only bugfixes mergeable. It's up to the release manager though, and as this PR is strictly an addition, and doesn't change any code, maybe it'll be merged. Just if it takes until after the 1.14 release, please don't be surprised either.

otavio commented:
> @otavio, Just to set the expectations right, Zephyr is in the code freeze stage now for the 1.14 release, with officially only bugfixes mergeable. It's up to the release manager though, and as this PR is strictly an addition, and doesn't change any code, maybe it'll be merged. Just if it takes until after the 1.14 release, please don't be surprised either.

Absolutely.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,817 @@
+
+	k_sem_take(&ctx.semaphore, NETWORK_TIMEOUT * 2);
+
+	if (net_addr_pton(addr.sin_family, ip, &addr.sin_addr)) {
```
BTW, how well you have tested this part. What would happen if DNS resolving fails, then the ip variable will contain some garbage. The current code is trusting that `net_addr_pton()` fails in this case but it might be that it will not fail (if the ip address contains some numeric values, and yes, this is not very likely). What about initializing the ip variable to some value that will cause the `net_add_pton()` fail for sure, like placing there a non-numeric value. Then if DNS fails, we will catch that properly.

chtavares592 replied:
when I inform an ip that doesn't exist like test12345, the dns will fail to resolve and the function start coap will continuous running. The app will fail on the send request when the app call the send function. I will think about the way to improve that error and this fail early.

chtavares592 replied:
I understood now what you told me, and already fix it ;)

Commit:
lib: updatehub: add suport a CoAPS/DTLS

This extends the UpdateHub library code to allow the
use of CoAPS/DTLS for communication.

Refs: #13039.

Signed-off-by: Christian Tavares <christian.tavares@ossystems.com.br>

Commit:
lib: updatehub: add support ipv6

This extends the UpdateHub library code to allow the
use of IPV6 for communication.

Signed-off-by: Christian Tavares <christian.tavares@ossystems.com.br>

jukkar reviewed (approved):
LGTM

otavio commented:
Hello everyone (@Laczen @mike-scott @nvlsianpu),

We did address all the known issues and we'd like to ask if you guys could review it one more time.

Thanks in advance,

Laczen reviewed (approved):
Hi @otavio, I did not find anything (nvs/settings) that I could review.

mike-scott commented on lib/updatehub/shell.c:
```diff
@@ -0,0 +1,89 @@
+	char *firmware_version = k_malloc(BOOT_IMG_VER_STRLEN_MAX);
+
+	updatehub_get_firmware_version(firmware_version);
+	updatehub_get_device_identity(device_id);
```
Can add 2nd buffer length param here and pass in `DEVICE_ID_SIZE`.

chtavares592 replied:
If your comment about ``lib/updatehub/device.h`` is in a "resolved" section, could I guess assume this comment was resolved too?

chtavares592 replied:
Sorry, I understood now!

mike-scott commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,807 @@
+					   .sin_port = htons(5683) };
+	char ip[MAX_IP_SIZE] = "UNKNOW_IP";
+
+	if (dns_get_addr_info(UPDATEHUB_SERVER, DNS_QUERY_TYPE_A, NULL,
```
This is purely a style comment, but there is a blocking DNS API in the socket layer:
https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/socket.h#L206
(Usable as `getaddrinfo()`)

Example usage in a sample: https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/big_http_download/src/big_http_download.c#L277

Seems you could simplify here and get future benefits by using that API.  @pfalcon can weigh in, but most other socket samples using DNS are using the socket-based API.

mike-scott reviewed (commented):
I didn't have time to run through the sample on my K64F -- but just a couple minor questions.

mike-scott commented:
NOTE: My comment in `lib/updatehub/device.h` is in a "resolved" section.

pfalcon commented:
> (Usable as getaddrinfo())

> Seems you could simplify here and get future benefits by using that API. @pfalcon can weigh in, but most other socket samples using DNS are using the socket-based API.

I'd personally recommend to use POSIX functions, and make samples POSIX compatible in general. Using native Zephyr functions could be used as an optimization, but IMHO ability to debug (large part of) update process on Linux outweights *premature* optimization. Of course, if it's a thought out/profiled move to use dns_get_addr_info(), sounds good.

mike-scott commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,38 @@
+	memset(id, 0, id_max_len);
+
+	for (i = 0; i < length; i++) {
+		snprintk(buf, id_max_len, "%02x", hwinfo_id[i]);
```
Using id_max_len here would allow for an overrun of buf which is defined as `char buf[2];` above.  And if we were to change this to `sizeof(buf)` as it probably should be, then a size of 2 doesn't leave room for a terminating NULL.  Recommend making the definition:
`char buf[3];`
and here use
`snprintk(buf, sizeof(buf), "%02x", hwinfo_id[i]);`

mike-scott commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,38 @@
+	for (i = 0; i < length; i++) {
+		snprintk(buf, id_max_len, "%02x", hwinfo_id[i]);
+		buf_len = buf_len + strlen(buf);
+		strncat(id, buf, MIN(id_max_len, buf_len));
```
The size value passed into strncat() represents the maximum # of chars to be appended.  So the logic is a bit wrong here.

We don't need to keep track of `buf_len`, because strlen(buf) is the maximum amount to be appended (2).   And the smallest amount to be appended would be the left over size available in id (because it's being added to each pass through the loop).

Recommend dropping use of `buf_len` and change `strncat()` line to something like:
`strncat(id, buf, MIN(id_max_len - strlen(id), strlen(buf)));`

Note: mentally compiled only.  You'll need to test that this works.

jukkar replied:
> Recommend dropping use of `buf_len` and change `strncat()` line to something like:
> `strncat(id, buf, MIN(id_max_len - strlen(id), strlen(buf)));`

We should avoid having function calls in MIN() or MAX() macros because how those macros are implemented. In worst case, the functions might be called multiple times which is usually not desired.

otavio replied:
So we are going to put extra variables to avoid this.

mike-scott commented:
Woops.. I didn't realize the CoAP / DTLS patch is not included in this PR.  I may have added some comments there.  I'm gtg with the 1st patch.

otavio commented:
@galak, considering that all requested changes on this PR have been solved, is it possible to get this merged on 1.14.0 and has been under work since Feb 4th?

I know the merge window is closed but this is very low risk and isolated. We are waiting for this PR to be merged to send the DTLS the IPv6 support PRs.

jukkar commented on samples/net/updatehub/README.rst:
```diff
@@ -81,6 +83,7 @@ Using this server the user need create own ``overaly-prj.conf`` setting the opti
 UpdateHub need too set the option :option:`CONFIG_UPDATEHUB_POLL_INTERVAL` with the period of
 your preference, remembering that the limit is between 0 minute until 43200 minutes(30 days).
+And this server does not use DTLS, so you do not need to add the ``overlay-dtls.config``.
 
```
BTW, could there be simple instructions how to create the .der files if user wants use own files instead of the supplied ones.

d3zd3z replied:
The docs should be really clear that the supplied keys should only ever be used for testing. Aside from this, since the private key is compiled into the image, it means that every device will require a separate image. Really, the key needs to be stored in some kind of protected storage on the device. This is one problem that TF-M seeks to address.

otavio replied:
@d3zd3z 
> The docs should be really clear that the supplied keys should only ever be used for testing.

Agreed.

> Aside from this, since the private key is compiled into the image, it means that every device will require a separate image. Really, the key needs to be stored in some kind of protected storage on the device. This is one problem that TF-M seeks to address.

Could you shed some light on the TF-M? Not sure I follow you here.

d3zd3z replied:
Zephyr doesn't currently have anywhere to store a secret, safely. This kind of API should come in some day, though, and TF-M is one solution that may provide that.

otavio replied:
Ok, in fact, for production I'd expect it to be written during the manufacture so it can be individual and not part of the firmware.

What do you think we should do now? Leave it as is?

galak commented on include/updatehub.h:
```diff
@@ -0,0 +1,77 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
```
We need a different location for this header, but not sure what the best location for it.

galak replied:
Will raise this in our Thur Dev-review meeting to get input on what the path should be.

otavio replied:
@galak did you figure where do you guys wish this to be moved?

I'd like to get this merged soon and this seems to be the last open item.

galak commented on include/updatehub.h:
```diff
@@ -0,0 +1,77 @@
+	UPDATEHUB_OK,
+	UPDATEHUB_HAS_UPDATE,
+	UPDATEHUB_NO_UPDATE,
+
```
remove blank line.

galak commented on lib/updatehub/Kconfig:
```diff
@@ -0,0 +1,88 @@
+	  Set the interval that the UpdateHub update server will be polled.
+	  This time interval is zero and 43200 minutes(30 days).
+
+config UPDATEHUB_PRODUCT_UID
```
Can this use the hwinfo (drivers/hwinfo) for UID?

chtavares592 replied:
No, because product-uid is about products on the UpdateHub Cloud, is not related to hardware.

galak commented on lib/updatehub/Kconfig:
```diff
@@ -0,0 +1,88 @@
+	  Allow the use of UpdateHub Community
+	  Server (updatehub-ce) as alternative to the
+	  updatehub.io enterprise server.
+
```
remove blank line.

galak commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,39 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DEVICE_H__
```
Change to __UPDATEHUB_DEVICE_H__ as __DEVICE_H__ might conflict.

jukkar replied:
It actually seems that include protector in this case should be named something like this `ZEPHYR_INCLUDE_UPDATEHUB_DEVICE_H_`, see existing header files in include dir for examples.

galak commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,39 @@
+
+#define DEVICE_ID_MAX_SIZE 65
+
+static inline bool updatehub_get_device_identity(char *id, int id_max_len)
```
Is there a reason this is in a header?

chtavares592 replied:
Because that function is used on shell.c and the updatehub.c.

d3zd3z replied:
Is there any reason to inline this, though? Why not just have a prototype in this file, and put the function body in a C file. The function call overhead is very small compared to what the function is doing, and inlining will only waste space.

galak replied:
I'm asking why in a .h vs .c?

chtavares592 replied:
Okay, I understood now.

galak commented on lib/updatehub/firmware.h:
```diff
@@ -0,0 +1,31 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __FIRMWARE_H__
```
Change to __UPDATEHUB_FIRMWARE_H__

galak commented on lib/updatehub/firmware.h:
```diff
@@ -0,0 +1,31 @@
+#include <dfu/mcuboot.h>
+#include <dfu/flash_img.h>
+
+static inline bool updatehub_get_firmware_version(char *version)
```
Same question, why is this in a header?

chtavares592 replied:
Same here

d3zd3z replied:
Again, put a prototype here, and put the code in a C file.

galak commented on lib/updatehub/shell.c:
```diff
@@ -0,0 +1,89 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_UPDATEHUB_SHELL)
```
This if defined() should not be needed, should be handled by CMakefile.

galak reviewed (dismissed):
Some cleanups needed.

Need to figure out where to place the header.

d3zd3z commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,39 @@
+
+	for (i = 0; i < length; i++) {
+		snprintk(buf, sizeof(buf), "%02x", hwinfo_id[i]);
+		id_len = strlen(id);
```
Calling `strlen` and even `strncat` within a loop will result in quadratic performance.

d3zd3z commented on lib/updatehub/device.h:
```diff
@@ -0,0 +1,39 @@
+		snprintk(buf, sizeof(buf), "%02x", hwinfo_id[i]);
+		id_len = strlen(id);
+		buf_len = strlen(buf);
+		strncat(id, buf, MIN(id_max_len - id_len, buf_len));
```
This shouldn't be necessary. `strncat` will stop when it reaches the end of `src`, so it is adequate to just pass this as `strncat(id, buf, id_max_len - id_len)`. That also helps the behavior, since the `buf_len` will always be 2.

d3zd3z commented on lib/updatehub/updatehub_priv.h:
```diff
@@ -0,0 +1,221 @@
+	UPDATEHUB_STATE_ERROR,
+};
+
+static char *updatehub_response(enum updatehub_response response)
```
Also, do these need to be in the header? I'm not seeing a performance reason to inline them, and regardless they aren't marked for inline anyway. This is going to cause a separate copy of this function to be built into every C file that includes (and uses) this function.

chtavares592 replied:
This is used just on updatehub.c, I've separated because I wanted to turn cleaner updatehub.c.
But, if is better, I can put this on updatehub.c.

d3zd3z commented on lib/updatehub/updatehub_priv.h:
```diff
@@ -0,0 +1,221 @@
+	struct updatehub_config_device_identity device_identity;
+};
+
+static const struct json_obj_descr recv_probe_objects_descr[] = {
```
These also should be in a C file, with just a prototype in the header. Otherwise, you are causing the data to be included multiple times.

chtavares592 replied:
This is used just on updatehub.c, I've separated because I wanted to turn cleaner updatehub.c.
But, if is better, I can put this on updatehub.c.

d3zd3z commented on samples/net/updatehub/src/c_certificates.h:
```diff
@@ -0,0 +1,21 @@
+
+#define CA_CERTIFICATE_TAG 1
+
+static const unsigned char server_certificate[] = {
```
Again, why is this in a header file? The code and data should be in C files.

chtavares592 replied:
I followed an example provided here for testing with dtls, in this example (https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/sockets/echo_client/src) the certificates were used in .h file, so I thought this was the default template for setting certificates.

d3zd3z replied:
I'm not sure why that sample is in a header file, either. I'm just thinking from a code point of view. You should assume a header file will be included in multiple places, otherwise there really isn't much point in putting it in a header file.

otavio replied:
Right but this needs to be consistent on the whole project. I am keen to keep this as is for now and we can fix it globally next. What do you think?

d3zd3z reviewed (dismissed):
Overall, this looks good. Any idea how resilient the update process is to a poor network? I haven't had a chance to do a deeper security analysis of the process itself, and will try to get some time to do that as well.

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,766 @@
+	char *firmware_version = k_malloc(BOOT_IMG_VER_STRLEN_MAX);
+
+	if (device_id == NULL || firmware_version == NULL) {
+		LOG_ERR("Could not alloc device_id or firmware_version memory");
```
possible memory leak here if only one of the allocations fail

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,766 @@
+
+	wait_fds();
+
+	int rcvd = recv(ctx.sock, data, MAX_DOWNLOAD_DATA, MSG_DONTWAIT);
```
Please introduce variables at the start of the block

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,766 @@
+		break;
+	}
+
+	if (strncmp(report.previous_state, "", strlen("")) != 0) {
```
Instead of strlen(""), use `sizeof("") - 1` in order to avoid function call

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,766 @@
+
+	wait_fds();
+
+	int rcvd = recv(ctx.sock, metadata, MAX_PAYLOAD_SIZE, MSG_DONTWAIT);
```
variable declarations to the start of func

jukkar commented on lib/updatehub/updatehub.c:
```diff
@@ -0,0 +1,766 @@
+
+	if (device_id == NULL || firmware_version == NULL ||
+	    metadata == NULL || metadata_copy == NULL) {
+		LOG_ERR("Could not alloc probe memory");
```
possible memory leak here

jukkar commented on lib/updatehub/updatehub_firmware.c:
```diff
@@ -0,0 +1,24 @@
+
+#include "updatehub_firmware.h"
+
+bool updatehub_get_firmware_version(char *version)
```
Better would be to put the version buffer length as a second parameter

otavio replied:
We can do that, however, we are using the `BOOT_IMG_VER_STRLEN_MAX` constant which is defined elsewhere.

jukkar replied:
Sure and that works for now, it is just that if someone calls this function but uses different length for the version, then invalid memory access might occur. If the length is part of the function call, that risk of that wold be minimized.

jukkar commented on lib/updatehub/updatehub_priv.h:
```diff
@@ -0,0 +1,221 @@
+	struct resp_probe_objects_array objects[2];
+	size_t objects_len;
+	const char *product;
+	const char *supported_hardware[40];
```
Instead of magic constant, please create a define that describes the value better. Why value 40 here, could the value be configured via Kconfig if end user wants to minimize the amount of memory allocated?

jukkar reviewed (approved):
LGTM

otavio commented:
@galak @d3zd3z do you mind to have a new look at this? We are not aware of any pending change.

otavio commented:
Hello everyone. We believe it is ready to go and we addressed all issues. Is it possible for you guys to check it?

carlescufi commented:
Looked through this and the only real unresolved question is where the updatehub.h header belongs. Since @nashif is doing a cleanup of headers anyway, I suggest we discuss where to place it after the fact in order to prevent this PR from bitrotting.

It could be in `include/dfu` or `include/mgmt`
